From f6c6b65a3dac511cc2ea83a32cb99b5e0ae362b9 Mon Sep 17 00:00:00 2001
From: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Date: Thu, 15 Feb 2024 19:15:54 +0100
Subject: [PATCH 1014/1695] ta: remoteproc: Add PTA call to get access right to
 the memories

Get access to the memories used by the remote processor to be able to
load the firmware images.

The memories are released after the load

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 7b1a4b4cb89fd7ebb08aec5b6024e4de3470402c ta: remoteproc: Add PTA call to get access right to the memories]
Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Change-Id: I1d64f78eab9f693b614b7c901e82ccfa28b05437
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/361064
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/391546
Tested-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 ta/remoteproc/src/remoteproc_core.c | 48 +++++++++++++++++++++++++++--
 1 file changed, 46 insertions(+), 2 deletions(-)

diff --git a/ta/remoteproc/src/remoteproc_core.c b/ta/remoteproc/src/remoteproc_core.c
index 11369a5f8..136f5f874 100644
--- a/ta/remoteproc/src/remoteproc_core.c
+++ b/ta/remoteproc/src/remoteproc_core.c
@@ -777,6 +777,36 @@ static void remoteproc_clean_resources(struct remoteproc_context *ctx)
 	assert(!res);
 }
 
+static TEE_Result remoteproc_get_mems(struct remoteproc_context *ctx)
+{
+	uint32_t param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+					       TEE_PARAM_TYPE_NONE,
+					       TEE_PARAM_TYPE_NONE,
+					       TEE_PARAM_TYPE_NONE);
+	TEE_Param params[TEE_NUM_PARAMS] = { };
+
+	params[0].value.a = ctx->rproc_id;
+
+	return TEE_InvokeTACommand(pta_session, TEE_TIMEOUT_INFINITE,
+				  PTA_REMOTEPROC_GET_MEM, param_types,
+				  params, NULL);
+}
+
+static TEE_Result remoteproc_release_mems(struct remoteproc_context *ctx)
+{
+	uint32_t param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+					       TEE_PARAM_TYPE_NONE,
+					       TEE_PARAM_TYPE_NONE,
+					       TEE_PARAM_TYPE_NONE);
+	TEE_Param params[TEE_NUM_PARAMS] = { };
+
+	params[0].value.a = ctx->rproc_id;
+
+	return TEE_InvokeTACommand(pta_session, TEE_TIMEOUT_INFINITE,
+				  PTA_REMOTEPROC_RELEASE_MEM, param_types,
+				  params, NULL);
+}
+
 static TEE_Result remoteproc_load_segment(uint8_t *src, uint32_t size,
 					  uint32_t da, uint32_t mem_size,
 					  void *priv)
@@ -953,6 +983,7 @@ static TEE_Result remoteproc_load_fw(uint32_t pt,
 	struct remoteproc_context *ctx = NULL;
 	uint32_t rproc_id = params[0].value.a;
 	TEE_Result res = TEE_ERROR_GENERIC;
+	TEE_Result res2 = TEE_ERROR_GENERIC;
 
 	if (pt != exp_pt)
 		return TEE_ERROR_BAD_PARAMETERS;
@@ -972,19 +1003,32 @@ static TEE_Result remoteproc_load_fw(uint32_t pt,
 	DMSG("Got base addr: %p size %#zx", params[1].memref.buffer,
 	     params[1].memref.size);
 
+	res = remoteproc_get_mems(ctx);
+	if (res) {
+		EMSG("Can't get access right to memories");
+		goto out;
+	}
+
 	res = remoteproc_verify_firmware(ctx, params[1].memref.buffer,
 					 params[1].memref.size);
 	if (res) {
 		EMSG("Can't Authenticate the firmware (res = %#"PRIx32")", res);
-		goto out;
+		goto put_mems;
 	}
 
 	res = remoteproc_load_elf(ctx);
 	if (res)
-		goto out;
+		goto put_mems;
 
 	ctx->state = REMOTEPROC_LOADED;
 
+put_mems:
+	res2 = remoteproc_release_mems(ctx);
+	if (res2 && ctx->state == REMOTEPROC_LOADED) {
+		remoteproc_clean_resources(ctx);
+		ctx->state = REMOTEPROC_OFF;
+		res = res2;
+	}
 out:
 	/* Clear reference to firmware image from shared memory */
 	ctx->fw_img = NULL;
-- 
2.47.2

